pkg/clusterapi/aws/actuator.go

```diff
@@ -395,49 +396,37 @@ func (a *Actuator) Update(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 		// but instance could be deleted between the two calls.
 		return fmt.Errorf("attempted to update machine but no instances found")
 	}
+	SortInstances(instances)
 
 	// In very unusual circumstances, there could be more than one machine running matching this
 	// machine name and cluster ID. In this scenario we will keep the newest, and delete all others.
-	var newestInstance *ec2.Instance
-	if len(instances) > 1 {
-
-		// Determine which instance is the newest:
-		for _, instance := range instances {
-			mLog.WithFields(log.Fields{
-				"instanceID": *instance.InstanceId,
-				"state":      *instance.State.Name,
-				"launchTime": *instance.LaunchTime,
-			}).Warn("found multiple instances for machine in pending/running state")
-
-			if newestInstance == nil {
-				newestInstance = instance
-				continue
-			}
-			newestInstance = chooseNewestInstance(newestInstance, instance)
-		}
+	newestInstance := instances[0]
+	mLog = mLog.WithField("instanceID", *newestInstance.InstanceId)
+	mLog.Debug("instance found")
 
-		// Cleanup all the older instances:
+	if len(instances) > 1 {
 		terminateInstanceIDs := []*string{}
-		for _, instance := range instances {
+		// Cleanup all older instances:
+		for _, instance := range instances[1:] {
 			if *instance.InstanceId != *newestInstance.InstanceId {
+				mLog.WithFields(log.Fields{
+					"instanceID": *instance.InstanceId,
+					"state":      *instance.State.Name,
+					"launchTime": *instance.LaunchTime,
+				}).Warn("cleaning up extraneous instance for machine")
 				terminateInstanceIDs = append(terminateInstanceIDs, instance.InstanceId)
 			}
 		}
-		err := a.terminateInstances(client, terminateInstanceIDs, mLog)
+
+		err = a.terminateInstances(client, terminateInstanceIDs, mLog)
```

Comment:
In both places where `terminateInstances` is called, the caller is creating a slice of IDs from a slice of instances. Can we let the callee create the slice of IDs instead and just pass the slice of instances?

Reply:
Good idea.